<a href="https://colab.research.google.com/github/SofiaMarchesi/Titanic/blob/main/Titanic_Machine_Learning_part3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [50]:
from google.colab import drive
import pandas as pd

drive.mount('/content/drive')
train = pd.read_csv('/content/drive/My Drive/Colab Notebooks/Datasets/train_cabin.csv')
test = pd.read_csv('/content/drive/My Drive/Colab Notebooks/Datasets/test_cabin.csv')


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [51]:
train1=train.copy()
test1 = test.copy()

In [52]:
test1 = test1.fillna(-1)

In [53]:
from sklearn.model_selection import train_test_split,cross_val_score,GridSearchCV
from sklearn.metrics import confusion_matrix,classification_report,accuracy_score

import matplotlib.pyplot as plt
import numpy as np

from sklearn.preprocessing import MinMaxScaler,StandardScaler
from sklearn.ensemble import RandomForestClassifier, ExtraTreesClassifier,VotingClassifier, AdaBoostClassifier,GradientBoostingClassifier
from sklearn.linear_model import LogisticRegression,RidgeClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier


The best model was GradientBoost and RandomForest
 Kaggle = 0.78468 top 10%
 (the last model) 

# Machine Learning with all features

# RandomForestClassifier

In [54]:
features = ['Age', 'Fare','Name_Capt','Name_Master','Name_Miss', 'Name_Mr', 'Name_Mrs', 'Name_Ms','Name_Sir', 'Pclass_A', 
            'Pclass_B', 'Pclass_C', 'Sex_female','Sex_male', 'Embarked_C', 'Embarked_S','Family size_Alone', 'Family size_Huge', 
            'Family size_Large','Family size_Small', 'Cabin_B','Cabin_X']
y = train1['Survived']
X= train1[features]
X_train, X_test, y_train, y_test = train_test_split(X,y,random_state=0,test_size=0.5)

In [55]:
rf = RandomForestClassifier(criterion= 'entropy', max_depth= 5,max_features= 6, min_samples_leaf= 4, n_estimators=150, random_state=0).fit(X_train,y_train)

rf_cross = cross_val_score(rf, X_train,y_train, cv=10, n_jobs=-1)

y_pred_rf1 =  rf.predict(X_test)

print('Cross validation Scores:\n', (rf_cross))
print('Cross validation Scores Média:\n', (rf_cross.mean()))
print('Score test {:.4f}'.format(rf.score(X_test,y_test)))
print('Score train: {:.4f}'.format(rf.score(X_train,y_train)))
print(confusion_matrix(y_test, y_pred_rf1))
print(classification_report(y_test, y_pred_rf1))

Cross validation Scores:
 [0.86666667 0.75555556 0.82222222 0.82222222 0.8        0.77272727
 0.84090909 0.88636364 0.84090909 0.84090909]
Cross validation Scores Média:
 0.8248484848484848
Score test 0.8296
Score train: 0.8472
[[248  35]
 [ 41 122]]
              precision    recall  f1-score   support

           0       0.86      0.88      0.87       283
           1       0.78      0.75      0.76       163

    accuracy                           0.83       446
   macro avg       0.82      0.81      0.81       446
weighted avg       0.83      0.83      0.83       446



- 0.8229 - max_depth = 6
- 0.8251 - max_depth = 5
- 0.8274 - max_feature = 6
- 0.8274 - min_samples_leaf = 4
- 0.8296 - n_estimators = 150


# SVC

In [56]:
y = train1['Survived']
X = StandardScaler().fit_transform(train1[features])
X_train, X_test, y_train, y_test = train_test_split(X,y,random_state=0,test_size=0.5)

In [41]:
'''param_grid= {'C':[0.00001,0.0001,0.001,0.01,0.1,1,5,10,20], 'kernel': ['linear', 'poly', 'rbf', 'sigmoid'],'degree': range(2,5),'gamma':['scale', 'auto'],'probability': [True,False],}


svc_cv = GridSearchCV(SVC(), param_grid, cv=10).fit(X_train,y_train)


# Print the tuned parameters and score
print("Tuned Decision Tree Parameters: {}".format(svc_cv.best_params_))
print("Best score is {}".format(svc_cv.best_score_))'''

'param_grid= {\'C\':[0.00001,0.0001,0.001,0.01,0.1,1,5,10,20], \'kernel\': [\'linear\', \'poly\', \'rbf\', \'sigmoid\'],\'degree\': range(2,5),\'gamma\':[\'scale\', \'auto\'],\'probability\': [True,False],}\n\n\nsvc_cv = GridSearchCV(SVC(), param_grid, cv=10).fit(X_train,y_train)\n\n\n# Print the tuned parameters and score\nprint("Tuned Decision Tree Parameters: {}".format(svc_cv.best_params_))\nprint("Best score is {}".format(svc_cv.best_score_))'

In [57]:
svc = SVC(C=1, degree = 2, gamma= 'scale', kernel = 'rbf', probability = True).fit(X_train,y_train)


svc_cross = cross_val_score(svc, X_train,y_train, cv=20, n_jobs=-1)

y_pred_svc =  svc.predict(X_test)

print('Cross validation Scores:\n', (svc_cross))
print('Cross validation Scores Média:\n', (svc_cross.mean()))
print('Score test {:.4f}'.format(svc.score(X_test,y_test)))
print('Score train: {:.4f}'.format(svc.score(X_train,y_train)))
print(confusion_matrix(y_test, y_pred_svc))
print(classification_report(y_test, y_pred_svc))

Cross validation Scores:
 [0.82608696 0.91304348 0.69565217 0.91304348 0.82608696 0.81818182
 0.90909091 0.81818182 0.72727273 0.81818182 0.72727273 0.77272727
 0.90909091 0.77272727 0.86363636 0.86363636 0.86363636 0.86363636
 0.90909091 0.86363636]
Cross validation Scores Média:
 0.8336956521739131
Score test 0.8206
Score train: 0.8449
[[248  35]
 [ 45 118]]
              precision    recall  f1-score   support

           0       0.85      0.88      0.86       283
           1       0.77      0.72      0.75       163

    accuracy                           0.82       446
   macro avg       0.81      0.80      0.80       446
weighted avg       0.82      0.82      0.82       446



# Machine learning with less features:
[features_new2]

In [58]:
features_new2 = ['Sex_female','Name_Mr','Sex_male','Fare','Age','Pclass_A','Pclass_C','Cabin_X','Name_Miss','Name_Mrs','Family size_Small',
                'Family size_Huge','Family size_Alone','Embarked_S','Pclass_B','Embarked_C','Name_Master','Cabin_C','Family size_Large',
                'Cabin_B','Embarked_Q','Cabin_E','Cabin_A','Cabin_D','Cabin_F']
y = train1['Survived']
X= train1[features_new2]
X_train, X_test, y_train, y_test = train_test_split(X,y,random_state=0,test_size=0.5)

Tuned Decision Tree Parameters: {'criterion': 'entropy', 'max_depth': 9, 'max_features': 3, 'min_samples_leaf': 2, 'n_estimators': 150} Best score is 0.8337078651685393

In [59]:
rf3 = RandomForestClassifier(criterion= 'entropy', max_depth= 9,max_features= 3, min_samples_leaf= 2, n_estimators= 150, random_state=0).fit(X_train,y_train)

rf3_cross = cross_val_score(rf3, X_train,y_train, cv=20, n_jobs=-1)

y_pred_rf3 =  rf3.predict(X_test)

print('Cross validation Scores:\n', (rf3_cross))
print('Cross validation Scores Média:\n', (rf3_cross.mean()))
print('Score test {:.4f}'.format(rf3.score(X_test,y_test)))
print('Score train: {:.4f}'.format(rf3.score(X_train,y_train)))
print(confusion_matrix(y_test, y_pred_rf3))
print(classification_report(y_test, y_pred_rf3))

Cross validation Scores:
 [0.82608696 0.86956522 0.69565217 0.86956522 0.82608696 0.77272727
 0.86363636 0.77272727 0.81818182 0.77272727 0.68181818 0.77272727
 0.86363636 0.77272727 0.86363636 0.86363636 0.90909091 0.86363636
 0.81818182 0.81818182]
Cross validation Scores Média:
 0.8157114624505928
Score test 0.8229
Score train: 0.8697
[[243  40]
 [ 39 124]]
              precision    recall  f1-score   support

           0       0.86      0.86      0.86       283
           1       0.76      0.76      0.76       163

    accuracy                           0.82       446
   macro avg       0.81      0.81      0.81       446
weighted avg       0.82      0.82      0.82       446



In [60]:
features_importances = pd.DataFrame({'feature': features_new2,'importances':rf3.feature_importances_})
features_order = features_importances.sort_values(by='importances',ascending = False)
print(features_order)

              feature  importances
1             Name_Mr     0.153748
4                 Age     0.128569
3                Fare     0.116469
2            Sex_male     0.105960
0          Sex_female     0.104638
6            Pclass_C     0.049361
7             Cabin_X     0.037669
5            Pclass_A     0.033865
8           Name_Miss     0.033766
10  Family size_Small     0.033132
9            Name_Mrs     0.031563
13         Embarked_S     0.021143
12  Family size_Alone     0.019725
15         Embarked_C     0.019064
11   Family size_Huge     0.018776
14           Pclass_B     0.016801
18  Family size_Large     0.014254
16        Name_Master     0.013445
21            Cabin_E     0.010259
19            Cabin_B     0.010101
17            Cabin_C     0.008555
20         Embarked_Q     0.007197
23            Cabin_D     0.005616
22            Cabin_A     0.004291
24            Cabin_F     0.002036


# RandomForest with less features
[features_new3]

In [61]:
features_new3 = ['Sex_female','Name_Mr','Sex_male','Fare','Age','Pclass_A','Pclass_C','Cabin_X','Name_Miss','Name_Mrs','Family size_Small',
                'Family size_Huge','Family size_Alone','Embarked_S','Pclass_B','Embarked_C','Name_Master','Cabin_C','Family size_Large',
                'Cabin_B','Cabin_E','Cabin_D']
y = train1['Survived']
X= train1[features_new3]
X_train, X_test, y_train, y_test = train_test_split(X,y,random_state=0,test_size=0.5)

In [42]:
# Setup the parameters and distributions to sample from: param_dist
'''param_dist = {"n_estimators": [150,200],
              "max_features": range(1,5),
              'max_depth': [4,5,6],
              "min_samples_leaf": [1,2,3,4,5,6],
              "criterion": ["gini", "entropy"]}


rt4 = RandomForestClassifier(n_jobs=-1)

rt4_cv = GridSearchCV(rt4, param_dist, cv=5, n_jobs=-1).fit(X_train,y_train)


# Print the tuned parameters and score
print("Tuned Decision Tree Parameters: {}".format(rt4_cv.best_params_))
print("Best score is {}".format(rt4_cv.best_score_)) '''

'param_dist = {"n_estimators": [150,200],\n              "max_features": range(1,5),\n              \'max_depth\': [4,5,6],\n              "min_samples_leaf": [1,2,3,4,5,6],\n              "criterion": ["gini", "entropy"]}\n\n\nrt4 = RandomForestClassifier(n_jobs=-1)\n\nrt4_cv = GridSearchCV(rt4, param_dist, cv=5, n_jobs=-1).fit(X_train,y_train)\n\n\n# Print the tuned parameters and score\nprint("Tuned Decision Tree Parameters: {}".format(rt4_cv.best_params_))\nprint("Best score is {}".format(rt4_cv.best_score_)) '

In [62]:
rf4 = RandomForestClassifier(criterion= 'gini', max_depth= 5,max_features= 1, min_samples_leaf= 1, n_estimators= 150, random_state=0).fit(X_train,y_train)

rf4_cross = cross_val_score(rf4, X_train,y_train, cv=20, n_jobs=-1)

y_pred_rf4 =  rf4.predict(X_test)

print('Cross validation Scores:\n', (rf4_cross))
print('Cross validation Scores Média:\n', (rf4_cross.mean()))
print('Score test {:.4f}'.format(rf4.score(X_test,y_test)))
print('Score train: {:.4f}'.format(rf4.score(X_train,y_train)))
print(confusion_matrix(y_test, y_pred_rf4))
print(classification_report(y_test, y_pred_rf4))

Cross validation Scores:
 [0.82608696 0.86956522 0.69565217 0.86956522 0.7826087  0.81818182
 0.86363636 0.77272727 0.81818182 0.81818182 0.68181818 0.77272727
 0.86363636 0.81818182 0.81818182 0.86363636 0.90909091 0.86363636
 0.81818182 0.86363636]
Cross validation Scores Média:
 0.8203557312252965
Score test 0.8139
Score train: 0.8629
[[239  44]
 [ 39 124]]
              precision    recall  f1-score   support

           0       0.86      0.84      0.85       283
           1       0.74      0.76      0.75       163

    accuracy                           0.81       446
   macro avg       0.80      0.80      0.80       446
weighted avg       0.82      0.81      0.81       446



In [63]:
features_importances = pd.DataFrame({'feature': features_new3,'importances':rf4.feature_importances_})
features_order = features_importances.sort_values(by='importances',ascending = False)
print(features_order)

              feature  importances
0          Sex_female     0.134327
1             Name_Mr     0.121842
2            Sex_male     0.118080
3                Fare     0.074582
4                 Age     0.070828
9            Name_Mrs     0.068313
7             Cabin_X     0.054281
5            Pclass_A     0.048989
6            Pclass_C     0.046259
10  Family size_Small     0.043958
8           Name_Miss     0.041414
15         Embarked_C     0.025704
13         Embarked_S     0.025174
12  Family size_Alone     0.023458
19            Cabin_B     0.020429
20            Cabin_E     0.020193
16        Name_Master     0.012571
11   Family size_Huge     0.011741
14           Pclass_B     0.010037
18  Family size_Large     0.009878
17            Cabin_C     0.009513
21            Cabin_D     0.008431


# RandomForest with GradientBoost with more features:
[features0]
- kaggle = 0.78468 top 10%

In [64]:
from sklearn.model_selection import train_test_split
features0 = ['Age', 'Fare', 'Name_Dona','Name_Capt', 'Name_Col', 'Name_Don', 'Name_Dr', 'Name_Jonkheer','Name_Lady', 'Name_Major', 'Name_Master',
            'Name_Miss', 'Name_Mlle','Name_Mme', 'Name_Mr', 'Name_Mrs', 'Name_Ms', 'Name_Rev', 'Name_Sir','Name_the Countess', 'Pclass_A', 
            'Pclass_B', 'Pclass_C', 'Sex_female','Sex_male', 'Embarked_C', 'Embarked_Q', 'Embarked_S','Family size_Alone', 'Family size_Huge', 
            'Family size_Large','Family size_Small', 'Cabin_A', 'Cabin_B', 'Cabin_C', 'Cabin_D','Cabin_E', 'Cabin_F', 'Cabin_G', 'Cabin_T', 
            'Cabin_X']
y = train1['Survived']
X= train1[features0]


X_train, X_test, y_train, y_test = train_test_split(X,y,random_state=0,test_size=0.5)

In [65]:
rfboo1 = RandomForestClassifier(criterion= 'entropy', max_depth= 6,max_features= 8, min_samples_leaf= 4, n_estimators= 150).fit(X_train,y_train)


In [45]:
'''param_dist = {'loss': ['desviance','exponential'],
              'learning_rate':np.linspace(0.1,0.0001,11),
              'n_estimators':np.linspace(50,400,15),
              'min_samples_split':range(2,10),
              'min_samples_leaf':range(1,10),
              'criterion':['friedman_mse', 'mse', 'mae'],
              'max_depth':range(1,10)}

gb1 = GradientBoostingClassifier(rfboo1)

gb1_cv = GridSearchCV(gb1, param_dist, cv=5, n_jobs=-1).fit(X_train,y_train)'''



"param_dist = {'loss': ['desviance','exponential'],\n              'learning_rate':np.linspace(0.1,0.0001,11),\n              'n_estimators':np.linspace(50,400,15),\n              'min_samples_split':range(2,10),\n              'min_samples_leaf':range(1,10),\n              'criterion':['friedman_mse', 'mse', 'mae'],\n              'max_depth':range(1,10)}\n\ngb1 = GradientBoostingClassifier(rfboo1)\n\ngb1_cv = GridSearchCV(gb1, param_dist, cv=5, n_jobs=-1).fit(X_train,y_train)"

In [66]:
param_dist = {'loss': ['exponential'], 'learning_rate':[0.0175,0.018,0.0185,0.0184],'min_samples_split':[2],'min_samples_leaf':[1],
              'max_depth':[4]}

gb1 = GradientBoostingClassifier(rfboo1)

gb1_cv = GridSearchCV(gb1, param_dist, cv=5, n_jobs=-1).fit(X_train,y_train)

In [67]:
print("Tuned Decision Tree Parameters: {}".format(gb1_cv.best_params_))
print("Best score is {}".format(gb1_cv.best_score_))

Tuned Decision Tree Parameters: {'learning_rate': 0.018, 'loss': 'exponential', 'max_depth': 4, 'min_samples_leaf': 1, 'min_samples_split': 2}
Best score is 0.8337078651685393


Tuned Decision Tree Parameters: {'learning_rate': 0.018, 'loss': 'exponential', 'max_depth': 4, 'min_samples_leaf': 1, 'min_samples_split': 2}
Best score is 0.8337078651685393

In [68]:
rfboo1 = RandomForestClassifier(criterion= 'entropy', max_depth= 6,max_features= 8, min_samples_leaf= 4, n_estimators= 150).fit(X_train,y_train)

In [69]:
gb2 = GradientBoostingClassifier(learning_rate=0.018, loss= 'exponential', max_depth= 4, min_samples_leaf= 1, min_samples_split= 2).fit(X_train,y_train)

gb2_cross = cross_val_score(gb2,X_train,y_train, cv=20, n_jobs=-1)
y_pred_gb2 = gb2.predict(X_test)

print('Cross validation Scores:\n', (gb2_cross))
print('Cross validation Scores Média:\n', (gb2_cross.mean()))
print('Score test {:.4f}'.format(gb2.score(X_test,y_test)))
print('Score train: {:.4f}'.format(gb2.score(X_train,y_train)))
print(confusion_matrix(y_test, y_pred_gb2))
print(classification_report(y_test, y_pred_gb2))


Cross validation Scores:
 [0.82608696 0.82608696 0.69565217 0.82608696 0.82608696 0.86363636
 0.86363636 0.81818182 0.81818182 0.77272727 0.72727273 0.68181818
 0.81818182 0.77272727 0.90909091 0.90909091 0.86363636 0.86363636
 0.86363636 0.81818182]
Cross validation Scores Média:
 0.818181818181818
Score test 0.8229
Score train: 0.8764
[[248  35]
 [ 44 119]]
              precision    recall  f1-score   support

           0       0.85      0.88      0.86       283
           1       0.77      0.73      0.75       163

    accuracy                           0.82       446
   macro avg       0.81      0.80      0.81       446
weighted avg       0.82      0.82      0.82       446



Tuned Decision Tree Parameters: {'learning_rate': 0.0085, 'loss': 'exponential', 'max_depth': 4, 'min_samples_leaf': 2, 'min_samples_split': 3}
Best score is 0.8314606741573034

In [ ]:
param_dist = {'loss': ['desviance','exponential'], 'learning_rate': [0.0085,0.0084,0.0083,0.0082],'min_samples_split':range(2,8),'min_samples_leaf':range(1,8),
              'max_depth':range(1,8)}

gb3 = GradientBoostingClassifier(random_state=0)

gb3_cv = GridSearchCV(gb1, param_dist, cv=10, n_jobs=-1).fit(X_train,y_train)

In [ ]:
print("Tuned Decision Tree Parameters: {}".format(gb3_cv.best_params_))
print("Best score is {}".format(gb3_cv.best_score_))

Tuned Decision Tree Parameters: {'learning_rate': 0.0085, 'loss': 'exponential', 'max_depth': 4, 'min_samples_leaf': 2, 'min_samples_split': 3}
Best score is 0.8314606741573034


In [ ]:
gb3 = GradientBoostingClassifier(learning_rate=0.0085, loss= 'exponential', max_depth= 4, min_samples_leaf= 2, min_samples_split= 3).fit(X_train,y_train)

gb3_cross = cross_val_score(gb3,X_train,y_train, cv=20, n_jobs=-1)
y_pred_gb3 = gb3.predict(X_test)

print('Cross validation Scores:\n', (gb3_cross))
print('Cross validation Scores Média:\n', (gb3_cross.mean()))
print('Score test {:.4f}'.format(gb3.score(X_test,y_test)))
print('Score train: {:.4f}'.format(gb3.score(X_train,y_train)))
print(confusion_matrix(y_test, y_pred_gb3))
print(classification_report(y_test, y_pred_gb3))

Cross validation Scores:
 [0.86956522 0.82608696 0.60869565 0.82608696 0.86956522 0.90909091
 0.86363636 0.72727273 0.86363636 0.77272727 0.77272727 0.63636364
 0.72727273 0.77272727 0.90909091 0.90909091 0.86363636 0.81818182
 0.90909091 0.86363636]
Cross validation Scores Média:
 0.8159090909090908
Score test 0.8229
Score train: 0.8674
[[256  27]
 [ 52 111]]
              precision    recall  f1-score   support

           0       0.83      0.90      0.87       283
           1       0.80      0.68      0.74       163

    accuracy                           0.82       446
   macro avg       0.82      0.79      0.80       446
weighted avg       0.82      0.82      0.82       446



In [ ]:
X_true = test1[features0]
X = train1[features0]
y= train1['Survived']

gb3_true = GradientBoostingClassifier(learning_rate=0.0085, loss= 'exponential', max_depth= 4, min_samples_leaf= 2, min_samples_split= 3).fit(X,y)

y_pred_gb3_true = gb3.predict(X_true)

In [ ]:
output = pd.DataFrame({'PassengerId': test1.PassengerId, 'Survived': y_pred_gb3_true})
output.to_csv('my_submission[gradientboost1].csv', index=False)
output

PassengerId  Survived
0            892         0
1            893         0
2            894         0
3            895         0
4            896         1
..           ...       ...
413         1305         0
414         1306         1
415         1307         0
416         1308         0
417         1309         1

[418 rows x 2 columns]